### ISSUE TITLE
Help/Advice with Electron Intergration

### ISSUE THREAD

**Issue by syrys** [mgechev/angular-seed#1449]
Hey All,

Im trying to intergrate electron to this project. Yes im aware that there exists the advance seed [here](https://github.com/NathanWalker/angular2-seed-advanced) that is based on this project, but it has too many integrations i do not need and has a structure im not comfortable with. I personally like the setup of this project better, thus attempting to intergrate electron directly here. My problem here is likely related to angular (or this project), so if you have a sec, i would appreciate any insight.

### Setup
So, creating an electron app is fairly simple (just trying to make a proof of concept). First get an html/js project (lets use the `build.dev` output of this seed project), then add a simple package.json file to the `/dist/dev` (you can steal [this one](https://github.com/NathanWalker/angular2-seed-advanced/blob/master/src/client/package.json)), then also add a js file that drives the electron app (you can steal a compiled version of [this one](https://github.com/NathanWalker/angular2-seed-advanced/blob/master/src/client/main.desktop.ts), theres a bit of cleanup you need to do though). All set? Sort of.

make sure you install electron dependencies to your main project, then run `"node_modules/.bin/electron" "./dist/dev"`. You might get some errors. Fix them up by removing `/` from the `index.html`s `<base href="">`, link to files correctly etc. You may also want to put a copy of the required dependencies from node_modules to `/dist/dev/node_modules`. Heck, you could copy the entire node_modules folder there if you are lazy.

### Problem
Once all the above are setup correctly, you run the app, and you see the following error (i added the console/debugtools to the js file): http://i.imgur.com/IAxGsd2.jpg

````
EXCEPTION: Uncaught (in promise): Error: Cannot match any routes: 'E%3A/projects/angular2-seed/dist/dev'
...
Unhandled Promise rejection: Cannot match any routes: 'E%3A/projects/angular2-seed/dist/dev' ; Zone: angular ; Task: Promise.then ; Value: Error: Cannot match any routes: 'E%3A/projects/angular2-seed/dist/dev'(…) Error: Cannot match any routes: 'E%3A/projects/angular2-seed/dist/dev'
````

As you can see, the project loads. If i click about tab, you see the content. If you click home tab again you see the content (data.json doesnt get loaded, but thats fine, that can be fixed in `system-config.js`). But on the first load, as shown on the screenshot, you get a blank page with the nav bar. Seems like angular gets confused about routing. It feels like this is some obvious little thing im missing (due to my lack of understanding of this seed). Does anyone see any obvious problems that might point me in the right direction?

Also, im curious. What is the purpose of `system-config.js`?

TY in advance,
Cheers.

**Comment by mgechev** [comment]
I haven't tried to develop an electron application with the seed so my advices may not be very valuable to you. Given the philosophy of the project, I'm not sure you're taking the most maintainable direction here. I wouldn't recommend you to modify anything in `dist` since it contains assets which are produced by the build and are being removed during it's lifecycle.

I'd recommend you to create an electron-related tasks in `tools/tasks/project` similar to what @NathanWalker [did here](https://github.com/NathanWalker/angular2-seed-advanced/blob/45b32df180c28236ef72b649c79f47519e2ed99b/tools/tasks/project/desktop.package.windows.ts).

In general, I'm not sure where the problem may come from but if you can navigate through the pages and you have the issue only during the initial page load, you can try to set a default route.

**Comment by syrys** [comment]
@mgechev Thanks for the response.
Yeah ofcorse, i wont be using `/dist` going forward, thats silly haha. Im just copying files there as a proof of concept to get the app running (so i dont have to compile every little code change i make). Ofcorse in test setup i have them compiling through gulp tasks, but that would have taken too long to explain. I explained the bare minimum required to get upto speed on what the end result of where i was (so anyone else can sort of understand/recreate).

Sadly besides the code coming from within the `/src/client`, rest of my code (proof of concept) looks fairly similar to @NathanWalker's (sorry to ping you). My main problem is, i dont know enough about this projects setup (i dont have a good understanding of how core-js, systemjs, rxjs, zonejs works and the purpose of this system-config.js file, im ok with angular2, so i can follow the code but i dont understand the base setup just yet). So im suspecting that there is something related to one of those libraries that would be incharge of routing is causing some issue. Im slowly doing more and more comparisons with NathanWalkers project to see what tweaks he has done, so maybe i might come across the key difference. I will try mess with the default route in the meantime, thats a good suggestion.

If anyone can figure out what im missing from the error msg on the screenshot earlier, any help/suggestions are greatly appreciated. :D

**Comment by mgechev** [comment]
`system-config.js` is the place where we place the SystemJS development configuration and parts of it are overridden for unit tests. You need core-js for polyfilling missing ES6 (ES2015) APIs, SystemJS is the module loader which we use in development. Basically we are using all the tools the way they are supposed to be used so you can take a look at their documentation in order to get better understanding.

Unfortunately with the current development ecosystem there are a lot of moving parts, which is exactly the purpose of the seed - to reduce the learning gap (at least initially) and provide a quick, reliable and flexible start. I'd recommend you to take what @NathanWalker has build in the advanced seed and start removing parts until you get a minimum working Electron setup. After that you can see what the difference is with angular-seed and you will be sure what you did wrong, it might not be trivial and it may take a bit :-).

Good luck! Let me know how it goes.

**Comment by syrys** [comment]
@mgechev thanks for the breakdown of the libraries. that helps a lot, specially knowing that they are being used the way they are supposed to. I will definitely spend some time looking them up and doing my own learning too.

As for your suggestion on stripping down NathanWalkers project, i have been doing the opposite (building this project up towards his project), but i suppose your suggestion makes sense (specially given the other path didnt work). That was on my todo list anyway, but i will bump it up in priority a bit. :D 

Yup, i will try to post an update here if i manage to get it up and running. 

Basically my plan is to make a seed that builds a web app and a desktop app (modular in a way that its easy to just do one and not the other, and not add any more functionality besides that). Also integration with phonegap if i feel like it (i prefer phonegap over nativescript). Any other features or libraries are really upto the end user, just give them clear instructions on how to integrate them. Also i liked separate `/src/client` code from `/src/electron` code. Ontop of that, an easy was to add npm libraries to use with electron (which is really confusing on the other project). Lets see how i go :D

**Comment by syrys** [comment]
@mgechev 
After a bit of debugging i figured it out. it was a bit obvious really. Since electron apps run by calling the index.html by file not by domain/host like a web server, you cant do html5 routing (dah). so essentially (if your project was at `E:/projects/angular2-seed/dist/dev/index.html`) a route of `/about` would try look for a file `file:///E:/about` which would generate an error (although it would navigate within angular2, it will fail to fetch that file obviously, no big deal but i dont like errors). Default route wont work either unless you setup some fallback 404 route. Simply by setting angular2 to use hash routing you fix all that. 

hash routing steup looks something like this: `routerModule = RouterModule.forRoot(routes, {useHash: true});`... do a search for some of that text within the NathanWalkers advance seed to figure out where it goes.

I do have couple of minor errors i need to get rid of, but those should be easy enough to debug. So, hopefully, anyone else wanting to setup electron with this project, this topic should give all the info required to do so.

If i come across any other dead ends, i will post here, and later chuck it in a github project and post the link here for you guys to share around. @mgechev great seed project btw 👍
